File: scraper.py
import requests
from bs4 import BeautifulSoup
import re
import time
from urllib.parse import quote_plus

class MovieScraper:
    def __init__(self):
        self.session = requests.Session()
        self.session.headers.update({
            'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
        })
    
    def search_yts(self, query, year=None):
        """Scrape YTS.mx for magnet links (example)"""
        search_url = f"https://yts.mx/browse-movies/{quote_plus(query)}/all/all/0/latest"
        
        try:
            resp = self.session.get(search_url, timeout=15)
            soup = BeautifulSoup(resp.text, 'lxml')
            
            # Find first movie result
            movie_div = soup.find('div', class_='movie-tile')
            if not movie_div:
                return None
            
            # Get download page
            link = movie_div.find('a', href=True)
            if not link:
                return None
            
            detail_url = "https://yts.mx" + link['href']
            detail_resp = self.session.get(detail_url, timeout=15)
            detail_soup = BeautifulSoup(detail_resp.text, 'lxml')
            
            # Extract magnet for 720p
            magnet_tag = detail_soup.find('a', href=re.compile(r'magnet:?'))
            if magnet_tag:
                return {
                    'title': link.get('title', query),
                    'magnet': magnet_tag['href'],
                    'quality': '720p'
                }
        except Exception as e:
            print(f"Scraper error: {e}")
        
        return None
    
    def search_1337x(self, query):
        """Scrape 1337x.to for torrents"""
        search_url = f"https://1337x.to/search/{quote_plus(query)}/1/"
        
        try:
            resp = self.session.get(search_url, timeout=15)
            soup = BeautifulSoup(resp.text, 'lxml')
            
            # Get first result
            row = soup.find('tr')
            if not row:
                return None
            
            name_tag = row.find('a', class_='name')
            if not name_tag:
                return None
            
            magnet_link = row.find('a', href=re.compile(r'magnet:?'))
            if magnet_link:
                return {
                    'title': name_tag.text.strip(),
                    'magnet': magnet_link['href'],
                    'quality': '1080p'
                }
        except Exception as e:
            print(f"1337x error: {e}")
        
        return None
    
    def get_download_link(self, query):
        """Try multiple sources"""
        # Try YTS first
        result = self.search_yts(query)
        if result:
            return result
        
        # Try 1337x as fallback
        result = self.search_1337x(query)
        if result:
            return result
        
        return None

File: scraper.js
const axios = require('axios');
const cheerio = require('cheerio');

class MovieScraper {
  constructor() {
    this.client = axios.create({
      headers: {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
      },
      timeout: 15000
    });
  }

  async searchYts(query) {
    try {
      const searchUrl = `https://yts.mx/browse-movies/${encodeURIComponent(query)}/all/all/0/latest`;
      const response = await this.client.get(searchUrl);
      const $ = cheerio.load(response.data);

      const movieTile = $('.movie-tile').first();
      if (!movieTile.length) return null;

      const link = movieTile.find('a').first();
      if (!link.attr('href')) return null;

      const detailUrl = `https://yts.mx${link.attr('href')}`;
      const detailResponse = await this.client.get(detailUrl);
      const $$ = cheerio.load(detailResponse.data);

      const magnetTag = $$('a[href^="magnet:?"]').first();
      if (magnetTag.length) {
        return {
          title: link.text().trim() || query,
          magnet: magnetTag.attr('href'),
          quality: '720p'
        };
      }
    } catch (error) {
      console.error('YTS error:', error.message);
    }
    return null;
  }

  async search1337x(query) {
    try {
      const searchUrl = `https://1337x.to/search/${encodeURIComponent(query)}/1/`;
      const response = await this.client.get(searchUrl);
      const $ = cheerio.load(response.data);

      const row = $('tr').first();
      if (!row.length) return null;

      const nameTag = row.find('a.name').first();
      if (!nameTag.length) return null;

      const magnetLink = row.find('a[href^="magnet:?"]').first();
      if (magnetLink.length) {
        return {
          title: nameTag.text().trim(),
          magnet: magnetLink.attr('href'),
          quality: '1080p'
        };
      }
    } catch (error) {
      console.error('1337x error:', error.message);
    }
    return null;
  }

  async getDownloadLink(query) {
    let result = await this.searchYts(query);
    if (result) return result;
    result = await this.search1337x(query);
    if (result) return result;
    return null;
  }
}

module.exports = { MovieScraper };
